aspnetcore/security/authentication/mfa.md

```diff
@@ -0,0 +1,562 @@
+
+### MFA FIDO2 or passwordless
+
+FIDO2 is:
+
+* The most secure way of achieving MFA.
+* The only MFA flow that protects against phishing attacks.
+
+At present, ASP.NET Core doesn't support FIDO2 directly. FIDO2 can be used for MFA or passwordless flows.
+
+Azure Active Directory provides support for FIDO2 and passwordless flows. For more information, see [Passwordless authentication options for Azure Active Directory](/azure/active-directory/authentication/concept-authentication-passwordless).
+
+You can implement ASP.NET Core with FIDO2 by using the following OSS FIDO2 implementation:
+
+[FIDO2 .NET library for FIDO2 / WebAuthn Attestation and Assertion using .NET](https://github.com/abergs/fido2-net-lib)
+
+### MFA SMS
+
+MFA with SMS increases security massively compared with password authentication (Single factor). However, using SMS as a second factor is no longer recommended. Too many known attack vectors exist for this type of implementation.
+
+## Configure MFA for administration pages using ASP.NET Core Identity
+
+MFA could be forced on users to access sensitive pages within an ASP.NET Core Identity app. This could be useful for apps where different levels of access exist for the different identities. For example, users might be able to view the profile data using a password login, but an administrator would be required to use MFA to access the administrative pages.
+
+### Extend the login with an MFA claim
+
+The demo code is setup using ASP.NET Core with Identity and Razor Pages. The `AddIdentity` method is used instead of `AddDefaultIdentity` one, so an `IUserClaimsPrincipalFactory` implementation can be used to add claims to the identity after a successful login.
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+	services.AddDbContext<ApplicationDbContext>(options =>
+		options.UseSqlite(
+			Configuration.GetConnectionString("DefaultConnection")));
+	
+	services.AddIdentity<IdentityUser, IdentityRole>(
+		options => options.SignIn.RequireConfirmedAccount = false)
+	 .AddEntityFrameworkStores<ApplicationDbContext>()
+	 .AddDefaultTokenProviders();
+
+	services.AddSingleton<IEmailSender, EmailSender>();
+	services.AddScoped<IUserClaimsPrincipalFactory<IdentityUser>, 
+		AdditionalUserClaimsPrincipalFactory>();
+
+	services.AddAuthorization(options =>
+	{
+		options.AddPolicy("TwoFactorEnabled",
+			x => x.RequireClaim("TwoFactorEnabled", "true")
+		);
+	});
+
+	services.AddRazorPages();
+}
+```
+
+The `AdditionalUserClaimsPrincipalFactory` class adds the `TwoFactorEnabled` claim to the user claims after a successful login. This is only added after a login. The value is read from the database. This is added here because the user should only access the higher protected view, if the identity has logged in with MFA. If the database view was read from the database directly instead of using the claim, it would be possible to access the view without MFA directly after activating the MFA.
```

Comment:
```suggestion
The `AdditionalUserClaimsPrincipalFactory` class adds the `TwoFactorEnabled` claim to the user claims only after a successful login. The claim's value is read from the database. The claim is added here because the user should only access the higher protected view if the identity has logged in with MFA. If the database view is read from the database directly instead of using the claim, it's possible to access the view without MFA directly after activating the MFA.
```

Reply:
Scott this seems to contradict "The claim's value is read from the database" versus "If the database view is read from the database directly instead of using the claim, it's possible to access the view without MFA directly after activating the MFA.". The second statement is correct, the first isn't.

Reply:
text updated